src/imars3dv2/filters/tilt.py

```diff
@@ -0,0 +1,301 @@
+    # calculate the relative shift
+    shift_val = calculate_shift(image0, image1)
+    # crop both image to same range that contains the object
+    # NOTE:
+    #   1. if the shift is less than a pixel, the rotation center is basically at
+    #      the center, therefore no need to crop
+    #   2. DO NOT use shift from scipy.ndimage as it will distort the image by
+    #      its implied interpolation.
+    if shift_val < -1.0:
+        img0_tmp = image0[:, : int(shift_val)]
+        img180_tmp = np.fliplr(image1)[:, -int(shift_val) :]
+    elif shift_val > 1.0:
+        img0_tmp = image0[:, int(shift_val) :]
+        img180_tmp = np.fliplr(image1)[:, : -int(shift_val)]
+    else:
+        img0_tmp = image0
+        img180_tmp = np.fliplr(image1)
+
+    # normalize image
+    img0_tmp = (img0_tmp - img0_tmp.min()) / (img0_tmp.max() - img0_tmp.min())
+    img180_tmp = (img180_tmp - img180_tmp.min()) / (img180_tmp.max() - img180_tmp.min())
+
+    # rotate
+    # if the rotation axis is tilted by 2 deg, we need to tilt the image back by -2 deg
+    img0_tmp = rotate(
+        img0_tmp,
+        -tilt,
+        mode="edge",
+        resize=True,
+        preserve_range=True,
+        order=1,  # use default bi-linear interpolation for rotation
+    )
+    # since 180 is flipped, tilting back -2 deg of the original img180 means tilting +2 deg
+    # of the flipped one
+    img180_tmp = rotate(
+        img180_tmp,
+        +tilt,
+        mode="edge",
+        resize=True,
+        preserve_range=True,
+        order=1,  # use default bi-linear interpolation for rotation
+    )
+
+    # p-norm
+    # NOTE: p3 norm is selected as it makes the error function more sensitive
+    #       around the minimum.
+    diff = np.abs(img0_tmp - img180_tmp)
+    err = np.power(diff, 3).sum() / (np.linalg.norm(img0_tmp) * np.linalg.norm(img180_tmp))
+    return err
+
+
+def calculate_tilt(
+    image0: np.ndarray,
+    image180: np.ndarray,
+    low_bound: float = -5.0,
+    high_bound: float = -5.0,
+) -> float:
+    """
+    Use optimization to find the in-plane tilt angle.
+
+    Parameters
+    ----------
+    @param image0:
+        The first image for comparison, which is often the radiograph taken at
+        omega (< 180 deg)
+    @param image180:
+        The second image for comparison, which is often the radiograph taken at
+        omega + 180 deg
+    @param low_bound:
+        The lower bound of the tilt angle search space
+    @param high_bound:
+        The upper bound of the tilt angle search space
+
+    Returns
+    -------
+    @return:
+        The in-plane tilt angle in degrees
+    """
+    # make the error function
+    err_func = partial(calculate_dissimilarity, image0=image0, image1=image180)
+    # use bounded uni-variable optimizer to locate the tilt angle that minimize
+    # the dissimilarity of the 180 deg pair
+    res = minimize_scalar(
+        err_func,
+        bounds=(low_bound, high_bound),
+    )
+    #
+    return res.x
```

Comment:
Would it be good to return both the angle and a goodness of fit (e.g. chi-sq)?

Reply:
I can make it return the whole fitting results (`rst`), but the tilt correction really only cares about the final tilt results.

Reply:
It is now returning the whole `OptimizeResults` object (https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html) , which contains both the fitting results, residual, number of iterations and whether the optimization is successful or not.  I believe this should be enough for future fine tuning.